wrapper/src/main/java/software/amazon/jdbc/plugin/federatedauth/FederatedAuthConnectionPlugin.java

```diff
@@ -0,0 +1,347 @@
+
+    String host = getHost(hostSpec, props);
+
+    int port = getPort(hostSpec, props);
+
+    final Region region = getRegion(host, props);
+
+    final String cacheKey = getCacheKey(
+        PropertyDefinition.USER.getString(props),
+        host,
+        port,
+        region);
+
+    final TokenInfo tokenInfo = tokenCache.get(cacheKey);
+
+    final boolean isCachedToken = tokenInfo != null && !tokenInfo.isExpired();
+
+    if (isCachedToken) {
+      LOGGER.finest(
+          () -> Messages.get(
+              "FederatedAuthConnectionPlugin.useCachedIamToken",
+              new Object[] {tokenInfo.getToken()}));
+      PropertyDefinition.PASSWORD.set(props, tokenInfo.getToken());
+    } else {
+      getSetAndCacheAuthenticationToken(hostSpec, props, region, cacheKey);
+    }
+
+    try {
+      return connectFunc.call();
+    } catch (final SQLException exception) {
+      getSetAndCacheAuthenticationToken(hostSpec, props, region, cacheKey);
+      return connectFunc.call();
+    } catch (final Exception exception) {
+      LOGGER.warning(
+          () -> Messages.get(
+              "FederatedAuthConnectionPlugin.unhandledException",
+              new Object[] {exception}));
+      throw new SQLException(exception);
+    }
+  }
+
+  private void getSetAndCacheAuthenticationToken(HostSpec hostSpec, Properties props, Region region, String cacheKey)
+      throws SQLException {
+    final int tokenExpirationSec = IAM_EXPIRATION.getInteger(props);
+    final Instant tokenExpiry = Instant.now().plus(tokenExpirationSec, ChronoUnit.SECONDS);
+    AwsCredentialsProvider credentialsProvider =
+        this.credentialsProviderFactory.getAwsCredentialsProvider(hostSpec.getHost(), region, props);
+    String token = generateAuthenticationToken(
+        props,
+        hostSpec.getHost(),
+        getPort(hostSpec, props),
+        region,
+        credentialsProvider);
+    LOGGER.finest(
+        () -> Messages.get(
+            "FederatedAuthConnectionPlugin.generatedNewIamToken",
+            new Object[] {token}));
+    PropertyDefinition.PASSWORD.set(props, token);
+    tokenCache.put(
+        cacheKey,
+        new TokenInfo(token, tokenExpiry));
+    try {
+      credentialsProviderFactory.close();
+    } catch (IOException e) {
+      // Ignore
+    }
+  }
+
+  private Region getRegion(final String hostname, final Properties props) throws SQLException {
+    final String iamRegion = IAM_REGION.getString(props);
+    if (!StringUtils.isNullOrEmpty(iamRegion)) {
+      return Region.of(iamRegion);
+    }
+
+    // Fallback to using host
+    // Get Region
+    final String rdsRegion = rdsUtils.getRdsRegion(hostname);
+
+    if (StringUtils.isNullOrEmpty(rdsRegion)) {
+      // Does not match Amazon's Hostname, throw exception
+      final String exceptionMessage = Messages.get(
+          "FederatedAuthConnectionPlugin.unsupportedHostname",
+          new Object[] {hostname});
+
+      LOGGER.fine(() -> exceptionMessage);
```

Comment:
Use lambda in this case gives no advantages. You can pass message as a parameter.
LOGGER.fine(exceptionMessage);